crates/red_knot_python_semantic/src/types.rs

```diff
@@ -5404,6 +5448,63 @@ impl<'db> TupleType<'db> {
     }
 }
 
+macro_rules! type_guard_type_is_impl {
+    ($type:ident, $struct:ident) => {
+        impl<'db> $struct<'db> {
+            pub fn unbound(db: &'db dyn Db, ty: Type<'db>) -> Type<'db> {
+                Type::$type(Self::new(db, ty, None))
+            }
+
+            pub fn bound(
+                db: &'db dyn Db,
+                ty: Type<'db>,
+                scope: ScopeId<'db>,
+                symbol: ScopedSymbolId,
+                name: String,
+            ) -> Type<'db> {
+                Type::$type(Self::new(db, ty, Some((scope, symbol, name))))
+            }
+
+            #[must_use]
+            pub fn bind(
+                self,
+                db: &'db dyn Db,
+                scope: ScopeId<'db>,
+                symbol: ScopedSymbolId,
+                name: String,
+            ) -> Type<'db> {
+                Self::bound(db, *self.ty(db), scope, symbol, name)
+            }
+
+            pub fn is_bound(&self, db: &'db dyn Db) -> bool {
+                self.symbol_info(db).is_some()
+            }
+
+            pub fn is_unbound(&self, db: &'db dyn Db) -> bool {
+                self.symbol_info(db).is_none()
+            }
+        }
+    };
+}
+
+#[salsa::interned]
+pub struct TypeGuardType<'db> {
```

Comment:
It seems that the two types are exactly identical except for the name. Could we use a shared type (similar to `Class`) or make this type only thin wrapper types around a shared type? It would avoid the need for a macro